Fixes #1283, kill sexp forward/backward copies deletion to clipboard

xfthhxk opened the pull request:
<!-- ❤️ Thanks for filing a Pull Request on Calva! You are contributing to a better Clojure coding experience. ❤️ -->
<!-- We use checklists in order to not forget about important lessons we and others have learnt along the way. -->

## What has Changed?
<!-- Introduce the change(s) briefly here. Consider explaining why a particular change was implemented the way it was. If you have considered alternative ways to introduce the change, please elaborate a bit on that as well. -->
- added function `copyToClipboard` in `paredit/extension.ts`
- `paredit.killSexpForward` and `paredit.killSexpBackward` copy the deleted text to the clipboard.


<!-- Tell us what Github issue(s) your PR is fixing. Consider creating the issue if need be. -->
Fixes #1283 

## My Calva PR Checklist
<!-- Strike out (using `~`) items that do not apply, as Github reports how many are not ticked. If you want to add checkboxes, please do. -->

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [ ] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [x] Added to or updated docs in this branch, if appropriate
- [ ] Tested the VSIX built from the PR (so, after you've submitted the PR). You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test. NB: *You need to sign in/up at Circle CI to find the _Artifacts_ tab.*
     - [ ] Tested the particular change
     - [ ] Figured if the change might have some side effects and tested those as well.
     - [ ] Smoke tested the extension as such.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
     - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
     - [ ] If I am fixing just part of the issue, I have just referenced it w/o any of the "fixes” keywords.
- [x] Created the issue I am fixing/addressing, if it was not present.

## The Calva Team PR Checklist:
<!-- Please read the list, since you'll get a better idea about what to expect by doing so. 😄 -->

Before merging we (at least one of us) have:

- [ ] Made sure the PR is directed at the `dev` branch (unless reasons).
- [ ] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and tested it there if so.
- [ ] Read the source changes.
- [ ] Given feedback and guidance on source changes, if needed. (Please consider noting extra nice stuff as well.)
- [ ] Tested the VSIX built from the PR (well, if this is a PR that changes the source code.)
     - [ ] Tested the particular change
     - [ ] Figured if the change might have some side effects and tested those as well.
     - [ ] Smoke tested the extension as such.
- [ ] If need be, had a chat within the team about particular changes.

Ping @pez, @bpringe

<!-- This is a nice book to read about the power of checklists: https://www.samuelthomasdavies.com/book-summaries/health-fitness/the-checklist-manifesto/ -->

xfthhxk committed:
Fixes #1283, kill sexp forward/backward copies deletion to clipboard

PEZ commented:
Hello! Many thanks for contributing! The code changes look nice and clean(*). I left some comments on the issue about what I think is needed for the feature to be implemented in a way that won't trip up non-Emacs people.

(*) Usually we don't allow unrelated changes in the same PR. [This change](https://github.com/BetterThanTomorrow/calva/pull/1284/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R2409) is unrelated, but it is small and makes sense and we don't have a ton of changes on that file right now, so let's keep the change in this PR.

xfthhxk committed:
copy to clipboard for other paredit kill commands

PEZ commented on CHANGELOG.md:
```diff
@@ -3,6 +3,7 @@
 Changes to Calva.
 
 ## [Unreleased]
+- Fix [paredit.killSexpForward should copy deleted code to clipboard](https://github.com/BetterThanTomorrow/calva/issues/1283)
```
```suggestion
- [Add setting for letting Paredit Kill commands copy the deleted code to the clipboard](https://github.com/BetterThanTomorrow/calva/issues/1283)
```

PEZ commented on docs/site/paredit.md:
```diff
@@ -128,6 +128,8 @@ You can choose to disable all default key bindings by configuring `calva.paredit
 
 You can relax how Paredit's shortcuts replace VS Code built in shortcuts a bit by setting `calva.paredit.hijackVSCodeDefaults` to `false`.
 
+You can have the kill commands copy the deleted code to the clipboard by setting `calva.paredit.killAlsoCutsToKeyboard` to `true`.
+
```
This doesn't really belong under the sub header of **About the Keyboard Shortcuts**.

Since this PR is to some extent obsoleting the **!!! Note** on line 81. I suggest moving that note to the bottom of the **Editing** commands and reword it (well, replace the text with the text you have here, which is excellent).

xfthhxk replied:
Moved it and updated the wording.

PEZ commented on src/paredit/extension.ts:
```diff
@@ -12,11 +12,30 @@ let onPareditKeyMapChangedEmitter = new EventEmitter<String>();
+ * @param doc
+ * @param range
+ */
+function copyToClipboard(doc: EditableDocument, [start, end]){
```
Nice! I suggest a bit clearer name:
```suggestion
function copyRangeToClipboard(doc: EditableDocument, [start, end]){
```

PEZ replied:
Obviously we also need to fix the call sites. 😄

PEZ commented on package.json:
```diff
@@ -606,6 +606,12 @@
+                    },
+                    "calva.paredit.killAlsoCutsToKeyboard": {
+                        "type": "boolean",
+                        "markdownDescription": "When enabled, copies the deleted code to the clipboard simulating the Emacs implementation.",
```
Emacs users might take this as if there is a kill-ring they can add to. Maybe add some note about that consecutive kills are not adding to the clipboard, but instead replace the clipboard content?

xfthhxk replied:
Updated. No longer references Emacs so should be less confusing to Emacs users.

PEZ reviewed (changes requested):
Many thanks! I have some suggestions and comments. I'm trying to use the Review tool to relay them, but I don't know how that works, so let me know if you don't see any comments.

xfthhxk committed:
address PR comments

xfthhxk commented:
@PEZ Thank-you for the helpful suggestions.  I've incorporated them.  Also noticed I'd inadvertently named things with a `Keyboard` suffix instead of the intended `Clipboard`, eg `calva.paredit.killAlsoCutsToKeyboard` instead of `calva.paredit.killAlsoCutsToClipboard`; so this is also fixed.

PEZ commented on docs/site/paredit.md:
```diff
@@ -110,6 +107,11 @@ Default keybinding                | Action | Description
  `ctrl+alt+shift+q`                        | **Wrap Around ""** | Wraps the current form, or selection, with double quotes. Inside strings it will quote the quotes. <br> ![](images/paredit/wrap-around-quotes.gif)
  `ctrl+alt+r`<br>`ctrl+alt+p`/`s`/`c`/`q`                        | **Rewrap** | Changes enclosing brackets of the current form to parens/square brackets/curlies/double quotes.. <br> ![](images/paredit/rewrap.gif)
 
+!!! Note
```
I've recently learnt that you can have custom **Note** headings.

```suggestion
!!! Note "Copy to Clipboard when killing text"
```

PEZ commented on docs/site/paredit.md:
```diff
@@ -110,6 +107,11 @@ Default keybinding                | Action | Description
  `ctrl+alt+r`<br>`ctrl+alt+p`/`s`/`c`/`q`                        | **Rewrap** | Changes enclosing brackets of the current form to parens/square brackets/curlies/double quotes.. <br> ![](images/paredit/rewrap.gif)
 
+!!! Note
+    Many implementations of Paredit have commands for copying/cutting while deleting. In Calva, you can have the *kill* commands copy the deleted code to the clipboard by setting `calva.paredit.killAlsoCutsToKeyboard` to `true`.  Alternatively, you may use the [selection commands](#selecting) and copy/cut once you have the selection.
```
Found a remaining `Keyboard` 😄 (Also suggesting a slightly more to-the-point note.)
```suggestion
    You can have the *kill* commands always copy the deleted code to the clipboard by setting `calva.paredit.killAlsoCutsToClipboard` to `true`.  If you want to do this more on-demand, you can kill text by using the [selection commands](#selecting) and then *Cut* once you have the selection.
```

xfthhxk replied:
:man_facepalming: lol thanks for finding.

PEZ reviewed (approved):
Two small nit-picks, with ready suggestions to commit, if you like them.

Pre-approving. 😄

xfthhxk committed:
Update docs/site/paredit.md

Co-authored-by: Peter Strömberg <pez@pezius.com>

xfthhxk committed:
Update docs/site/paredit.md

Co-authored-by: Peter Strömberg <pez@pezius.com>

PEZ commented:
Thanks for this! It has been a pleasure reviewing and it is a pleasure merging.
